lib/config/formatter.go

```diff
@@ -0,0 +1,325 @@
+	// Defaults to filePathAndLine() if unspecified
+	FormatCaller func() (caller string)
+	// InitFormat is the initial format of log configuration
+	InitFormat []string
+	// InputFormat is the parsed format of each log line
+	InputFormat []string
+}
+
+// DefaultLoggerFormat is the default format in which to display logs (i.e. log.format in teleport.yaml is empty)
+var DefaultLoggerFormat = []string{"level", "component", "message", "caller"}
+var re = regexp.MustCompile(`[a-z]*`)
+
+type writer struct {
+	bytes.Buffer
+}
+
+const (
+	noColor = -1
+	red     = 31
+	yellow  = 33
+	blue    = 36
+	gray    = 37
+)
+
+// CheckAndSetDefaults checks and sets log format configuration
+func (tf *textFormatter) CheckAndSetDefaults() error {
+	// set padding
+	if tf.ComponentPadding == 0 {
+		tf.ComponentPadding = trace.DefaultComponentPadding
+	}
+	// set caller
+	if tf.FormatCaller == nil {
+		tf.FormatCaller = formatCallerWithPathAndLine
+	}
+	// set log formatting
+	if tf.InitFormat == nil {
+		tf.InputFormat = DefaultLoggerFormat
+		return nil
+	}
+	// parse input
+	res, err := parseInputFormat(tf.InitFormat)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	tf.InputFormat = res
+	return nil
+}
+
+func parseInputFormat(formatInput []string) ([]string, error) {
+	var result []string
+	values := map[string]bool{"level": true, "component": true, "message": true, "caller": true, "timestamp": true}
+	var key string
+	for _, component := range formatInput {
+		key = re.FindString(component)
+		key = strings.TrimSpace(key)
+		if !values[key] {
+			return nil, trace.BadParameter("invalid log format key: %v", key)
+		}
+		result = append(result, key)
+	}
+	return result, nil
+}
+
+// Format formats each log line as confiured in teleport config file
+func (tf *textFormatter) Format(e *log.Entry) ([]byte, error) {
+	var data []byte
+	caller := tf.FormatCaller()
+	w := &writer{}
+
+	for _, match := range tf.InputFormat {
+		switch match {
+		case "level":
+			color := noColor
+			if tf.EnableColors {
+				switch e.Level {
+				case log.DebugLevel:
+					color = gray
+				case log.WarnLevel:
+					color = yellow
+				case log.ErrorLevel, log.FatalLevel, log.PanicLevel:
+					color = red
+				default:
+					color = blue
+				}
+			}
+			w.writeField(strings.ToUpper(padMax(e.Level.String(), trace.DefaultLevelPadding)), color)
+		case "component":
+			padding := trace.DefaultComponentPadding
+			if tf.ComponentPadding != 0 {
+				padding = tf.ComponentPadding
+			}
+			if w.Len() > 0 {
+				w.WriteByte(' ')
+			}
+			value := e.Data[trace.Component]
+			var component string
+			if reflect.ValueOf(value).IsValid() {
+				component = fmt.Sprintf("[%v]", value)
+			}
+			component = strings.ToUpper(padMax(component, padding))
+			if component[len(component)-1] != ' ' {
+				component = component[:len(component)-1] + "]"
+			}
+			w.WriteString(component)
+		case "message":
+			if e.Message != "" {
+				w.writeField(e.Message, noColor)
+			}
+		case "caller":
```

Comment:
Like mentioned above - if either `timestamp` or `caller` has been specified - put them in the front/back as expected.